[PT2][Optimus] Fix a group batch fusion corner case

mengluy0125 opened the pull request:
Summary:
We have a user report on BA model that it raised "AttributeError: 'SymFloat' object has no attribute 'shape'", thus we add type check for the meta node.

See more context in the post
https://fb.workplace.com/groups/1075192433118967/permalink/1510477489590457/

Test Plan:
# local reproduce

```
CUDA_VISIBLE_DEVICES=3 OC_CAUSE=1 buck2 run mode/opt //scripts/jackiexu0313/pt2:local_model_with_pt2 -- --test_mode split-batch-decompose --flow_id 646303196
```

P1609807876

# E2E

before fix

f646303196

after fix

Differential Revision: D63399959


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

ezyang commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -396,10 +396,14 @@ def _pointwise_node_can_be_fused(self, node: torch.fx.Node):
+            and is_node_meta_valid(other)  # type: ignore[arg-type, union-attr]
+            # torch.SymInt or torch.SymFloat object has no attribute 'shape'
+            and hasattr(input.meta["val"], "shape")  # type: ignore[union-attr]
+            and hasattr(other.meta["val"], "shape")  # type: ignore[union-attr]
```
Wouldn't it be better to test if val is a torch.Tensor?

mengluy0125 replied:
Sure, I can check if it's FakeTensor

ezyang reviewed (approved):
btw Tensor is ok too

mengluy0125 committed:
[PT2][Optimus] Fix a group batch fusion corner case (#136650)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/136650

We have a user report on BA model that it raised "AttributeError: 'SymFloat' object has no attribute 'shape'", thus we add type check for the meta node.

See more context in the post
https://fb.workplace.com/groups/1075192433118967/permalink/1510477489590457/

Test Plan:
# local reproduce

```
CUDA_VISIBLE_DEVICES=3 OC_CAUSE=1 buck2 run mode/opt //scripts/jackiexu0313/pt2:local_model_with_pt2 -- --test_mode split-batch-decompose --flow_id 646303196
```

P1609807876

# E2E

before fix

f646303196

after fix

Reviewed By: ezyang

Differential Revision: D63399959
